rednotebook/gui/main_window.py

```diff
@@ -663,48 +668,80 @@ def show_message(self, title, msg, msg_type):
         else:
             self.statusbar.show_message(title, msg, msg_type)
 
+    def update_undo_redo_buttons(self, _gobj=None, force_disable=False):
+        if force_disable:
+            self.undo_action.set_sensitive(False)
+            self.redo_action.set_sensitive(False)
+        else:
+            can_undo = self.day_text_field.day_text_buffer.can_undo()
+            self.undo_action.set_sensitive(can_undo)
+            can_redo = self.day_text_field.day_text_buffer.can_redo()
+            self.redo_action.set_sensitive(can_redo)
+
 
 class DayEditor(editor.Editor):
+    n_recent_buffers = 10  # How many recent buffers to store
+    _t2t_highlighting = None
+    _style_scheme = None
+
     def __init__(self, *args, **kwargs):
         editor.Editor.__init__(self, *args, **kwargs)
         self.day = None
-        self.scrolled_win = self.day_text_view.get_parent()
+        # Store buffers for recently edited days - these preserve undo history
+        # and cursor position. Once a buffer drops out of this, it needs to be
+        # recreated: at this point, the cursor and undo are lost.
+        self.recent_buffers = OrderedDict()
+
+    def _get_t2t_highlighting(self):
+        if self._t2t_highlighting is None:
+            # Load our own copy of t2t syntax highlighting
+            lm = GtkSource.LanguageManager.get_default()
+            search_path = lm.get_search_path()
+            if filesystem.files_dir not in search_path:
+                search_path.insert(0, filesystem.files_dir)
+                lm.set_search_path(search_path)
+            self._t2t_highlighting = lm.get_language('t2t')
+        return self._t2t_highlighting
+
+    def _get_style_scheme(self):
+        if self._style_scheme is None:
+            # Load our customised variant of the Tango scheme
+            sm = GtkSource.StyleSchemeManager.get_default()
+            if filesystem.files_dir not in sm.get_search_path():
+                sm.prepend_search_path(filesystem.files_dir)
+            self._style_scheme = sm.get_scheme('rednotebook')
+        return self._style_scheme
+
+    def _get_buffer_for_day(self, day):
+        key = day.date
+        if key in self.recent_buffers:
+            self.recent_buffers.move_to_end(key)
+            return self.recent_buffers[key]
+
+        buf = self.recent_buffers[key] = \
+            GtkSource.Buffer.new_with_language(self._get_t2t_highlighting())
+        buf.set_style_scheme(self._get_style_scheme())
+        buf.begin_not_undoable_action()
+        buf.set_text(day.text)
+        buf.end_not_undoable_action()
+
+        while len(self.recent_buffers) > self.n_recent_buffers:
```

Comment:
Shouldn't an `if` instead of `while` suffice since only one buffer is added at a time?

Reply:
It should suffice; I was doing this to be defensive. If something else adds further buffers, this will always bring the length back to the limit defined as `n_recent_buffers`. I don't mind changing it if you prefer, though.

Reply:
OK, done.